server/src/com/cloud/network/vpc/VpcManagerImpl.java

```diff
@@ -809,15 +945,22 @@ public Vpc createVpc(final long zoneId, final long vpcOffId, final long vpcOwner
 
     @DB
     protected Vpc createVpc(final Boolean displayVpc, final VpcVO vpc) {
-        final String cidr = vpc.getCidr();
-        // Validate CIDR
-        if (!NetUtils.isValidCIDR(cidr)) {
-            throw new InvalidParameterValueException("Invalid CIDR specified " + cidr);
-        }
 
-        // cidr has to be RFC 1918 complient
-        if (!NetUtils.validateGuestCidr(cidr)) {
-            throw new InvalidParameterValueException("Guest Cidr " + cidr + " is not RFC1918 compliant");
+        try {
+            final CIDR cidr = CIDRFactory.getCIDR(vpc.getCidr());
+
+            final CIDR[] superCidr = getZoneSuperCidrList(vpc.zoneId);
+
+            if (_vpcOffServiceDao.areServicesSupportedByNetworkOffering(vpc.getVpcOfferingId(), Service.VPCDynamicRouting)) {
+                if (!NetUtils.validateGuestCidrForOSPF(cidr, superCidr)) {
```

Comment:
Can this validation be moved to the ``CIDR`` interface to support IPv4/IPv6 dual stack operation?